Clean up BinaryReader; rename to BinaryReaderDelegate

binji opened the pull request:
This removes the use of setjmp/longjmp as well.

sbc100 commented on src/binary-reader.cc:
```diff
@@ -206,70 +140,207 @@ size_t read_i32_leb128(const uint8_t* p,
+}
+
+Result WABT_PRINTF_FORMAT(2, 3) BinaryReader::OnError(const char* format, ...) {
+  WABT_SNPRINTF_ALLOCA(buffer, length, format);
```
This is really just nit/suggestion.

Could this method be better named?  OnError looks like a delegate method name (and indeed it is one).  All of the "return OnError()" read a little funny to me too.  There doesn't seem to be any point in the return value here either since its always Result::Error.   Would it be clearer if the calling code pattern was more like?:

```
ShowError(....)   // retuns void
return Result::Error;
```

ShowError?  or RaiseError?  Or HandleError?

binji replied:
Yeah, in other places I've done
```
PrintError(...)
return Result::Error;
```

I thought this new way was a little nicer since it is a one-liner, but it's less clear. I'll change it.

binji commented:
The callgraph didn't change, setjmp/longjmp was used to throw exceptions, but now I just bubble the errors up the stack.

binji commented:
OK, renamed `OnError` -> `PrintError`. Also switched some `if (...) { PrintError(...); return` to `ERROR_UNLESS(...)`. Take a look at https://github.com/WebAssembly/wabt/pull/451/commits/9442c76916701f72f0630f5533562dde59090620.

binji commented:
Sorry!
/me races to land first... :-P
